uvicorn/protocols/websockets/auto.py

```diff
@@ -1,10 +1,18 @@
+import logging
+
 try:
     import websockets  # noqa
 except ImportError:  # pragma: no cover
     try:
         import wsproto  # noqa
     except ImportError:
         AutoWebSocketsProtocol = None
+        logger = logging.getLogger("uvicorn.error")
+        logger.warning(
```

Comment:
Doesn't this warning _always_ trigger when using a minimal install?

I believe the default WS mode is "auto", meaning that we import this "auto protocol", which will be None and trigger this warning by default...

I'm actually not sure if/how we can check for "your app is using WebSocket, but dependencies aren't available" before a WebSocket request comes in. Thoughts?

Reply:
yes you nailed it, the ws flag is on auto by default, so `AutoWebSocketsProtocol` will be None, and we will get a warning set here if we try a handshake:
https://github.com/encode/uvicorn/blob/fa914bc8b3af38017f2438770616cab46f600141/uvicorn/protocols/http/h11_impl.py#L267

which sucks.

Reply:
> So what we can do instead is log this warning when the first WebSocket connection request is received. Maybe that means we do need to define the "auto" protocol in this case as a "null protocol" class that logs this warning? I'm not sure how this None protocol is used by asyncio.

like in the `config.load()` ? if ws_protocol_class is None > warning ?

Reply:
the current scenario is this:
- uvicorn minimal > auto ws > ws=None > auto http > h11 > handshake in h11 > `msg = "Unsupported upgrade request." `

The present PR is preemptively raising a warning:
- uvicorn minimal > auto ws > ws=None > `Warning blablabla ws cant be used`

We could also add this warning  to the existing h11 warning:
- uvicorn minimal > auto ws > ws=None > auto http > h11 > handshake in h11 > `msg = "Unsupported upgrade request. Warning blablabla ws cant be used" `

I just find it more deceptive when you want to use ws, but you'll have a useful info in both cases.

It's also true raising a warning for everyone when they dont want to use ws is probably annoying.

I have no preference, I let you choose ! :)

Reply:
ok after trying to put the warning in the httptools_impl.py I find it weird.
the scenario is very deceptive for a user, and on top of that for a regular user using ws and having a real upgrade issue he'll have the warning, so I prefer to put it here in the import logic